hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RSRpcServices.java

```diff
@@ -972,11 +1002,44 @@ private void doBatchOp(final RegionActionResult.Builder builder, final HRegion r
 
       OperationStatus[] codes = region.batchMutate(mArray, atomic, HConstants.NO_NONCE,
         HConstants.NO_NONCE);
+      if (atomic) {
+        LinkedList<ResultOrException> resultOrExceptions = new LinkedList<>();
```

Comment:
Better to use ArrayList instead of LinkedList. ArrayList is faster than LinkedList for most cases. I can not recall the exact number but at least for thousands elements ArrayList is still faster.

Reply:
The reason why I use LinkedList here is that I need to add a element at the beginning of the list as follows:
https://github.com/apache/hbase/pull/2630/files#diff-e4052cd5a1f1c93375e3fbc931dc4df220deebc78c0d53fd2435b47fd04cd807R1019-R1020
https://github.com/apache/hbase/pull/2630/files#diff-e4052cd5a1f1c93375e3fbc931dc4df220deebc78c0d53fd2435b47fd04cd807R1031

Still ArrayList is faster in that case?

Reply:
Oh, if we want to add element at first then ArrayList is not suitable. But usually we could use other array based data structure such as ArrayDeque. Let me take a look at the code.